build: cargo-deny integration

Sparkenstein opened the pull request:
This pull request tries to solve the #5006 

Continuing the discussion from thread:
> Could you point out which crates are unlicensed

internal crates test_util and test_plugin are unlicensed according to cargo-check. attaching log of `cargo check` on license:
https://del.dog/dimerastoj.log

will push other changes to severity soon

bartlomieju commented:
> internal crates test_util and test_plugin are unlicensed according to cargo-check. attaching log of cargo check on license:
https://del.dog/dimerastoj.log

Thanks @Sparkenstein, we should add `MIT` to the `license` field in these crates. It's quite surprising that `ring` is missing a license though.

bartlomieju commented on deny.toml:
```diff
@@ -11,11 +11,11 @@ vulnerability = "deny"
 # The lint level for unmaintained crates
 unmaintained = "warn"
```
Should be `deny`

bartlomieju commented on deny.toml:
```diff
@@ -30,16 +30,32 @@ confidence-threshold = 0.8
 [bans]
 # Lint level for when multiple versions of the same crate are detected
-multiple-versions = "warn"
+multiple-versions = "deny"
```
This one should probably be `warn` - we're aware there are multiple versions of some crates but unfortunately can't do anything about it right now.

Sparkenstein commented:
In this commit:
 - added CC0-1.0 for `notify` crate
 - allowed OpenSSL license
 - added MIT for test util and test plugin
 - added exception for `ring` crate
 - hardened all the checks

Now as you see in the CI, because of `bans` checks are now on `deny`, the CI is failing because there are
multiple duplicate crates. and I think it is fine. the thing is we can skip duplicate version checks for these crates in config but
the list will be too long to maintain. I was thinking if we should skip the `bans` check altogether? it checks only for duplicate crates and banned crates which we don't have any as of now at least.

bartlomieju commented on deny.toml:
```diff
@@ -0,0 +1,76 @@
+# Lint level for when multiple versions of the same crate are detected
+multiple-versions = "warn"
+# Lint level for when a crate version requirement is `*`
+wildcards = "allow"
```
This setting should be `deny`

bartlomieju commented:
Looks like we got two problematic dependencies: `rand 0.6.3` coming from `multipart 0.17.0` (no update available to the latter) and `net 0.2` coming from `mio 0.6` (which could be fixed by upgrading to Tokio 0.3, which is in progress in #7988)

bartlomieju commented:
@Sparkenstein we did update our dependencies to latest versions recently, could you please rebase the PR? I'm still interested in landing it.

ry commented:
I'm worried about the complexity this adds to CI. What if we just had a script we could run once in a while?

I don't think it's possible to completely abstract away the cost of adding dependencies. All dependencies to Deno must be reviewed and considered - part of that review is license check. If this auto check was completely free, it'd be fine, but it's adding more dependencies, and almost 100 lines of code that have to be maintained. I'm not sure it's worth it.

bartlomieju commented:
@Sparkenstein thank you for the PR, but after more consideration I agree with Ryan. Additionally I'm worried that we might be getting sudden failures on CI after bumping dependency versions in similar case as with `ring` when you opened a PR.

Thanks again, but respectfully closing without merge.
